client/signal-generator.go

```diff
@@ -12,129 +14,283 @@ import (
 
 // SignalGenerator config
 type SignalGenerator struct {
-	ID          string  `node:"id"`
-	Parent      string  `node:"parent"`
-	Description string  `point:"description"`
-	Frequency   float64 `point:"frequency"`
-	Amplitude   float64 `point:"amplitude"`
-	Offset      float64 `point:"offset"`
-	SampleRate  float64 `point:"sampleRate"`
-	Value       float64 `point:"value"`
-	Units       string  `point:"units"`
-	HighRate    bool    `point:"highRate"`
-	BatchPeriod int     `point:"batchPeriod"`
-	Disable     bool    `point:"disable"`
+	ID          string `node:"id"`
+	Parent      string `node:"parent"`
+	Description string `point:"description"`
+	Disable     bool   `point:"disable"`
+	Units       string `point:"units"`
+	// SignalType must be one of: "sine", "square", "triangle", or "random walk"
+	SignalType string  `point:"signalType"`
+	MinValue   float64 `point:"minValue"`
+	MaxValue   float64 `point:"maxValue"`
+	// InitialValue is the starting value for the signal generator.
+	// For random walk, this must be between MinValue and MaxValue. For wave
+	// functions, this must be in radians (i.e. between 0 and 2 * Pi).
+	InitialValue float64 `point:"initialValue"`
+	RoundTo      float64 `point:"roundTo"`
+	// SampleRate in Hz.
+	SampleRate float64 `point:"sampleRate"`
+	// HighRate flag indicates that the points should be emitted on the NATS
+	// subject for high-rate points
+	HighRate bool `point:"highRate"`
+	// BatchPeriod is the batch timer interval in ms. When the timer fires, it
+	// generates a batch of points at the specified SampleRate. If not set,
+	// timer will fire for each sample at SampleRate.
+	BatchPeriod int `point:"batchPeriod"`
+	// Frequency for wave functions (in Hz.)
+	Frequency float64 `point:"frequency"`
+	// Min./Max. increment amount for random walk function
+	MinIncrement float64 `point:"minIncrement"`
+	MaxIncrement float64 `point:"maxIncrement"`
+	// Current value
+	Value float64 `point:"value"`
 }
 
+/* TODO: Optimization
+
+Note that future designs may keep track of all running SignalGeneratorClients
+and manage only a single batch period timer that runs at a scheduled time.
+One way to do this is to keep track of each clients' next scheduled run time in
+a sorted list. Then, we schedule the timer to run at the soonest scheduled time.
+At that time, we process all clients' batches of points and reschedule another
+timer.
+
```

Comment:
Timers are a very useful concept in rule logic as well, so perhaps a common timer node type could be leveraged for this.